docs/frontend/testing/writing-playwright-tests-for-milmove.md

```diff
@@ -0,0 +1,132 @@
+to more easily discover and reuse your helper.
+
+### Wait for the page to load
+After performing an action which results in data on the page changing,
+use async assertions (see below) to wait for the page to be in the
+appropriate state. As much as possible, avoid tying your test to
+implementation details about the [network
+events](https://playwright.dev/docs/network#network-events) and
+instead use async assertions (see below) to wait until the data you
+need is visible on the page before proceeding.
+
+See more at the [playwright docs for navigation](https://playwright.dev/docs/navigations)
+
+### Prefer user-facing attributes instead of css selectors 
+Quoting from the [playwright docs on
+locators](https://playwright.dev/docs/locators#locating-elements)
+
+> Playwright comes with multiple built-in locators. To make tests
+> resilient, we recommend prioritizing user-facing attributes and
+> explicit contracts such as page.getByRole().
+
+### Avoid force clicking at (almost) all costs
+Force clicking bypasses the
+[actionability](https://playwright.dev/docs/actionability) checks and
+increases the chances of flaky tests.
+
+Instead of using a locator for the `input`, try clicking on the
+associated text or parent element. The experience with playwright so
+far on MilMove is that if you need to force click, it's probably a
+smell that there's something wrong with how the page is working (e.g.
+a component with a label of the empty string instead of
+`&nbsp;` ).
+
+### Use async assertions
+
+This is referenced in the [playwright assertions
+docs](https://playwright.dev/docs/writing-tests#assertions) and will
+be expounded upon here.
+
+Instead of doing
+```javascript
+  const editCount = await page.getByText('Edit').count();
+  expect(editCount).toEqual(2);
+```
+Use the async assertions which will wait until the condition is met,
+which makes tests much less flaky!
+```javascript
+  await expect(page.getByText('Edit')).toHaveCount(2);
+```
+### Accessing Traces in CircleCI
+Unfortunately, [playwright traces do not load in
+CircleCI](https://github.com/microsoft/playwright/issues/18108). The
+workaround is to download them to your developer machine and then run
+
+```shell
+playwright show-trace path/to/trace.zip
+```
+
+Please note that if you use Safari to download the trace file, it will
+automatically unzip it for you, which then confuses playwright.
+Download it with another browser like Chrome or zip it back up
+
+```shell
+(cd ~/Downloads && zip -r trace.zip randomhextracegoeshere) && playwright show-trace ~/Downloads/trace.zip
+```
+
+## Testharness in the Support API
+To make it easy for a test to create the data it needs on demand, a
+testharness in the support API has been created on the backend. The
+`utils/testharness.js` has all of the methods.
+
+If you need to create a new method, make sure you do not hard code
+object IDs so that the same test can be run in the same database
+multiple times in a row without error.
```

Comment:
**question**: If we add new endpoints for this, is there anything specific we need to keep in mind (other than the hard-coded IDs)? Or will that process be essentially the same as adding other endpoints?

Reply:
Once we get the initial PR merged I can link to more documentation, but adding new testharness generators does not require creating a new endpoint, just adding an entry to map.

Reply:
Oh ok, sounds good, thanks!